### PR TITLE
Change logic of mitohifi/mitohifi api

### PR DESCRIPTION
**Pull request by ksenia-krasheninnikova**
Input is specified directly by mitohifi -c/-r options via ext.args2

## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by muffato** [comment]
@ksenia-krasheninnikova . It now works ! I just added a Singularity option to mount `$HOME`.

You can ignore the `prettier` error as it comes from pre-existing files. By the way, I undid your merge from master because it was bringing more errors.

**Comment by muffato** [comment]
@scorreard @mahesh-panchal . You may be interested in this MitoHiFi module change. Would you be able to review it ?

File: modules/nf-core/mitohifi/mitohifi/main.nf [hunk 1]
```diff
@@ -41,11 +41,12 @@ process MITOHIFI_MITOHIFI {
...
-    def run_type = reads ? "-r ${reads}" :
-                    contigs ? "-c ${contigs}" :
-                    exit("Reads or contigs must be specified")
+    def args2 = task.ext.args2 ?: ''
+    def script_input = args2.equals("-r") ? "-r ${input}" :
+                args2.equals("-c") ? "-c ${input}" :
+                exit("-r for reads or -c for contigs must be specified")
     """
-    mitohifi.py ${run_type} -f ${ref_fa} -g ${ref_gb} -o ${mito_code} -t $task.cpus ${args}
+    mitohifi.py ${script_input} -f ${ref_fa} -g ${ref_gb} -o ${mito_code} -t $task.cpus ${args}
```
Comment (scorreard) [inline modules/nf-core/mitohifi/mitohifi/main.nf:49]:
```suggestion
    mitohifi.py ${script_input} \\
        -f ${ref_fa} \\
        -g ${ref_gb} \\
        -o ${mito_code} \\
        -t $task.cpus ${args}
```

File: tests/modules/nf-core/mitohifi/mitohifi/nextflow.config [hunk 2]
```diff
@@ -1,5 +1,7 @@
 process {
-
+    ext.args2 = '-c'
```
Comment (scorreard) [inline tests/modules/nf-core/mitohifi/mitohifi/nextflow.config:2]: Can you do a test with -r too or is that too complicated?

**Comment by scorreard** [comment]
It is indeed a great improvement! I did a couple of suggestions, but other than that it looks good.
Can you be more specific about the reason prettier fail? What previous files?

**Comment by muffato** [comment]
Prettier complains about some Conda environment.yml files that have been introduced recently at the hackathon. I noticed in this PR and another one that merging `master` in was in fact aggravating the situation by bringing a ton of failures in unrelated modules. That's why I'd rather wait for things to be solved on the master branch

File: tests/modules/nf-core/mitohifi/mitohifi/nextflow.config (hunk 2 above)
Comment (ksenia-krasheninnikova) [reply tests/modules/nf-core/mitohifi/mitohifi/nextflow.config]: Yes, we can do it. Thank you for suggestion.

**Comment by ksenia-krasheninnikova** [comment]
@muffato I've added another test for raw reads as @scorreard suggested. Github CI now evokes 182 check runs. I wonder if it happens because this branch is out of sync with the master. It also runs 'py-test changes' instead of 'pytest' and the run checks lack reports about the two mitohifi pytest runs. However both entry points are being picked up when I run `nf-core modules test mitohifi/mitohifi` locally.

**Review by scorreard — APPROVED** [review]
LGTM! Well done!!

**Review by muffato — APPROVED** [review]

**Comment by ksenia-krasheninnikova** [comment]
Thank you both! @muffato @scorreard

**Comment by mahesh-panchal** [comment]
Sorry, I was away.
However, doesn't this go against nf-core guidelines? The `-r` or `-c` is mandatory, and so must be an `input:` channel.
Also the `exit` is going to kill the workflow too no?

**Comment by muffato** [comment]
Thanks @mahesh-panchal . I've opened #4248 to fix the `exit 1`.

I didn't think of getting the `-c`/`-r` through a channel. The guidelines say _SHOULD_, not _MUST_ 😉  :
> Non-file mandatory arguments or arguments needed to modify the command to make the module run with no error, SHOULD be provided as value channels (for example lib_type in [salmon/quant](https://github.com/nf-core/modules/blob/master/modules/nf-core/salmon/quant/main.nf)) - see ‘Input/output options’ below.
>
> Input channel `val` declarations SHOULD be defined for all mandatory non-file inputs that are essential for the functioning of the tool (e.g. parameters, flags etc).
> - Mandatory non-file inputs are options that the tool MUST have to be able to be run.
> - These non-file inputs are typically booleans or strings, and must be documented as such in the corresponding entry in the `meta.yaml`.

but I think using a separate channel would indeed be better here:

1. It'd make it absolutely clear that the option must be passed from the [documentation](https://nf-co.re/modules/mitohifi_mitohifi) and the input channel structure.
2. The `-c`/`-r` mode is very tightly coupled to the data that flow around the module that it makes sense to have it defined it right in the sub-workflow rather than in `modules.config`. It also mitigates the risk that someone overrides it by accident.

@ksenia-krasheninnikova . Could you do that in another PR ?

**Pull request merged**